src/main/java/com/icthh/xm/tmf/ms/activation/service/FinishTransactionStrategy.java

```diff
@@ -0,0 +1,43 @@
+package com.icthh.xm.tmf.ms.activation.service;
+
+import com.icthh.xm.tmf.ms.activation.domain.SagaLog;
+import com.icthh.xm.tmf.ms.activation.domain.SagaTransaction;
+import com.icthh.xm.tmf.ms.activation.domain.spec.SagaTaskSpec;
+import com.icthh.xm.tmf.ms.activation.domain.spec.SagaTransactionSpec;
+import com.icthh.xm.tmf.ms.activation.repository.SagaLogRepository;
+import com.icthh.xm.tmf.ms.activation.repository.SagaTransactionRepository;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.icthh.xm.tmf.ms.activation.domain.SagaTransactionState.FINISHED;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+@Slf4j
+@RequiredArgsConstructor
+public class FinishTransactionStrategy implements TransactionStatusStrategy {
+
+    private final SagaTaskExecutor taskExecutor;
+    private final SagaTransactionRepository transactionRepository;
+    private final SagaLogRepository logRepository;
+
+    public void updateTransactionStatus(SagaTransaction transaction, SagaTransactionSpec transactionSpec,
+                                        Map<String, Object> taskContext) {
+        if (isAllTaskFinished(transaction, transactionSpec)) {
+            transactionRepository.save(transaction.setSagaTransactionState(FINISHED));
+            taskExecutor.onFinish(transaction, taskContext);
+        }
+    }
+
+    protected boolean isAllTaskFinished(SagaTransaction transaction, SagaTransactionSpec transactionSpec) {
```

Comment:
May be it is better, to evaluate if all is finished on DB side? So we will not load SagaLog on service side?

Reply:
Oh it's was just refactor of existing logic. What db query you suggest?

Reply:
```
WITH txTasks AS (
  SELECT key
  FROM saga_task_spec
  WHERE id IN (SELECT UNNEST(ARRAY[...]::text[])) -- Replace [...] with the list of task IDs from txTasks
)
, finished AS (
  SELECT *
  FROM saga_log
  WHERE transaction_id = :transactionId
    AND event_type_key IN (SELECT key FROM txTasks)
)
SELECT
  CASE WHEN COUNT(DISTINCT event_type_key) = COUNT(DISTINCT key) THEN true
       ELSE false
  END AS allTasksFinished
FROM txTasks
JOIN finished ON txTasks.key = finished.event_type_key;
```